File: 0206-reverse-linked-list/0206-reverse-linked-list.cpp
//Recursive approach
//Time Complexity O(n)
//Space Complexity O(n) due to recursion call stack
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if(head == nullptr || head->next == nullptr)
            return head;
        
        //reverse rest of list (recursive)
        ListNode* newHead = reverseList(head->next);

        //reverse logic
        head->next->next = head; //last node point back to head to complete
        head->next = nullptr; //prevent cycle

        return newHead;

    }
};

File: 0206-reverse-linked-list/0206-reverse-linked-list.java
//Iterative approach
//Time complexity - O(n)
//Space complexity - O(1) no new data structure introduce
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        //constraint allows 0 no. of nodes. so we need to have edge case to return if null or 1 node
        //loop while cur not null
        ////initialise a temp listnode to hold cur.next
        ////set the cur.next to prev --reversing pointer
        ///forward  prev pointer to cur
        ////forward cur pointer to temp
        //return prev
        ListNode prev =null;
        ListNode cur = head;
        if(head == null || head.next == null) return head;
        while(cur!=null){
            ListNode temp = cur.next;
            cur.next = prev;
            prev = cur;
            cur = temp;
        }
        return prev;

    }
}